### PR TITLE
wallet: optimize migration process, batch db transactions

### PR DESCRIPTION
**Pull request by furszy**
Last step in a chain of PRs (#26836, #28894, #28987, #29403).

#### Detailed Description:
The current wallet migration process performs only individual db writes. Accessing disk to
delete all legacy records, clone and clean each address book entry for every created wallet,
create each new descriptor (with their corresponding master key, caches and key pool), and
also clone and delete each transaction that requires to be transferred to a different wallet.

This work consolidates all individual disk writes into two batch operations. One for the descriptors
creation from the legacy data and a second one for the execution of the migration process itself.
Efficiently dumping all the information to disk at once atomically at the end of each process.

This represent a speed up and also a consistency improvement. During migration, we either
want to succeed or fail. No other outcomes should be accepted. We should never leave a
partially migrated wallet on disk and request the user to manually restore the previous wallet from
a backup (at least not if we can avoid it).

Since the speedup depends on the storage device, benchmark results can vary significantly.
Locally, I have seen a 15% speedup on a USB 3.2 pendrive.

#### Note for Testers:
The first commit introduces a benchmark for the migration process. This one can be
cherry-picked on top of master to compare results pre and post changes.

Please note that the benchmark setup may take some time (~70 seconds here) due to the absence
of a batching mechanism for the address generation process (`GetNewDestination()` calls).

### CODE REVIEW

**Review by pablomartin4btc — COMMENTED** [review]
Concept ACK.

I'll also perform the benchmark as instructed clearly in the description and will review the code later.

**Comment by theStack** [comment]
Concept ACK

**Comment by furszy** [comment]
Rebased after #28894 merge, next in the line is #28987.

**Review by furszy — COMMENTED** [review]
Rebased, conflicts solved. Next PR in the line is #28987.

**Comment by furszy** [comment]
Decoupled the address book cloning related commits to #26836.

**Comment by josibake** [comment]
Concept ACK

Starting to review, if you get a chance can you rebase this so that it's only the relevant outstanding commits?

**Review by furszy — COMMENTED** [review]
> Starting to review, if you get a chance can you rebase this so that it's only the relevant outstanding commits?

Done. Rebased and reworked part of it.
I still need to accommodate some code and improve the commit descriptions. But, at least the changes are reviewable now.

File: src/bench/wallet_migration.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            wallet->AddToWallet(MakeTransactionRef(mtx), TxStateInactive{}, /*update_wtx=*/nullptr, /*fFlushOnClose=*/false, /*rescanning_old_block=*/true);
+        }
+
+        // Unload so the migration process loads it
+        TestUnloadWallet(std::move(wallet));
+    }
+
+    int wallet_num = 0;
+    bench.epochs(NUM_WALLETS).run([&] {
+        std::string filename = strprintf("legacy_%d", wallet_num);
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/bench/wallet_migration.cpp:73]: > I don't like the c_str below. Also, it doesn't compile on Windows.

yeah, me neither.. I just did this 5 months ago. Fixing..

File: src/wallet/scriptpubkeyman.cpp [hunk 2]
```diff
@@ -1811,8 +1817,8 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
 
         // Make the DescriptorScriptPubKeyMan and get the scriptPubKeys
         auto desc_spk_man = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(m_storage, w_desc, m_keypool_size));
-        desc_spk_man->AddDescriptorKey(key, key.GetPubKey());
-        desc_spk_man->TopUp();
+        WITH_LOCK(desc_spk_man->cs_desc_man, desc_spk_man->AddDescriptorKeyWithDB(batch, key, key.GetPubKey()));
```
[the remark this answers is not in the document]
Comment (josibake) [reply src/wallet/scriptpubkeyman.cpp:1820]: Also, in `mapCryptedKey`, the `CPubKey` is already there, so you could pull it out and avoid recalculating it: https://github.com/bitcoin/bitcoin/pull/28574/commits/5cc89e5eba7330ea3ab64ca66e364170dcd0b4f7#diff-55ce403d9c3440b4a38a261b7d452f84dd9a1d09b2a5ef1b4d3ac0c3567b9106R1762

But I'm not sure that's worth the extra effort considering it would only make a difference if a majority of the keys come from `mapCryptedKey` and it seems that this isn't making much of an improvement anyways.

File: src/wallet/scriptpubkeyman.cpp [hunk 3]
```diff
@@ -1923,9 +1929,9 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
                 if (!GetKey(keyid, key)) {
                     continue;
                 }
-                desc_spk_man->AddDescriptorKey(key, key.GetPubKey());
+                WITH_LOCK(desc_spk_man->cs_desc_man, desc_spk_man->AddDescriptorKeyWithDB(batch, key, key.GetPubKey()));
```
Comment (josibake) [inline src/wallet/scriptpubkeyman.cpp:1932]: in "wallet: batch MigrateToDescriptor() db transactions" (https://github.com/bitcoin/bitcoin/pull/28574/commits/5cc89e5eba7330ea3ab64ca66e364170dcd0b4f7):

Could get rid of the `key.GetPubKey()` call here and completely get rid of the `privkeyid` set above:

https://github.com/bitcoin/bitcoin/pull/28574/commits/5cc89e5eba7330ea3ab64ca66e364170dcd0b4f7#diff-55ce403d9c3440b4a38a261b7d452f84dd9a1d09b2a5ef1b4d3ac0c3567b9106R1907-R1910

since `keys.origins` is a pair of the `keyid` and `CPubKey`. Running the bench with with this suggestion seemed to help:

```
# after
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|    6,395,421,573.00 |                0.16 |    0.0% |32,790,285,573.00 |11,895,327,005.00 |  2.757 |3,044,102,935.00 |    0.2% |      6.40 | `WalletMigration`
```

6.94 -> 6.84 -> 6.4

**Review by josibake — COMMENTED** [review]
Looking good! I noticed a lot of `key.GetPubKey()` that could be removed, left some suggestions and verified locally that they improved the bench. I guess one conceptual question is if we are doing the `GetPubKey()` calls to regenerate the pubkeys in case we don't trust the data in the legacy wallet?

**Review by furszy — COMMENTED** [review]
Thanks for the review josi!

File: src/wallet/scriptpubkeyman.cpp [hunk 4]
```diff
@@ -1782,6 +1782,12 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
         keyid_it++;
     }
 
+    WalletBatch batch(m_storage.GetDatabase());
+    if (!batch.TxnBegin()) {
+        LogPrintf("Error during descriptors migration, cannot initialize db transaction\n");
```
[the remark this answers is not in the document]
Comment (josibake) [reply src/wallet/scriptpubkeyman.cpp:1787]: I see, maybe "Error generating descriptors for migration, cannot initialize db transaction"? Also not a big deal, its just he current message is confusing to me in that we aren't migrating descriptors, but taking existing keys, creating descriptors and then saving them in a new wallet.

File: src/wallet/scriptpubkeyman.cpp (hunk 3 above)
Comment (josibake) [reply src/wallet/scriptpubkeyman.cpp:1932]: Makes sense! I misread the title as meaning this PR is both optimizing the process and batching db transactions, but its clear from the PR description that this is focused on just the batching logic.

File: src/wallet/scriptpubkeyman.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/scriptpubkeyman.cpp]: Sure. Done as suggested.

**Comment by josibake** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/28574/commits/868e2f64a9d7de3a980e64fd7acd7323214d3765

Code reviewed and verified the bench results. Thanks for taking the suggestions!

File: src/wallet/scriptpubkeyman.cpp [hunk 5]
```diff
@@ -2009,9 +2009,15 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
 }
 
 bool LegacyScriptPubKeyMan::DeleteRecords()
+{
+    return RunWithinTxn(m_storage.GetDatabase(), /*process_desc=*/"delete legacy records", [&](WalletBatch& batch){
+        return DeleteRecords(batch);
+    });
+}
```
Comment (achow101) [inline src/wallet/scriptpubkeyman.cpp:2016]: In cd694033173ec7de2a16187432a3d7aa33ff62ca "wallet: introduce 'DeleteRecords(batch)' to be used within an external db txn"

After this PR, this `DeleteRecords()` overload is only used by the tests, so I think it can just be removed (and perhaps never introduced).

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -4280,11 +4266,16 @@ bool DoMigration(CWallet& wallet, WalletContext& context, bilingual_str& error,
...
-        return false;
-    }
-    return true;
+    return RunWithinTxn(wallet.GetDatabase(), /*process_desc=*/"apply migration process", [&](WalletBatch& batch) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet){
+        if (auto res_migration = wallet.ApplyMigrationData(batch, *data); !res_migration) {
+            error = util::ErrorString(res_migration);
+            return false;
+        }
+        // Process succeed, connect the SPKM signals
+        wallet.ConnectScriptPubKeyManNotifiers();
```
Comment (achow101) [inline src/wallet/wallet.cpp:4275]: In 868e2f64a9d7de3a980e64fd7acd7323214d3765 "wallet: ApplyMigrationData, group all db txs into a single atomic operation"

I don't think it's necessary to connect the signals anymore since the local wallet exists outside of any contexts. Although removing this should be its own commit.

If you choose to not remove this, for completeness, this should also do `NotifyCanGetAddressesChanged` to match the previous behavior.

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -2334,12 +2334,22 @@ DBErrors CWallet::LoadWallet()
...
+    });
+    if (!res) return util::Error{util::ErrorString(res)};
+    if (!was_txn_committed) return util::Error{_("Error starting/committing db txn for wallet transactions removal process")};
 
+    // Now that txs have been deleted from disk, erase them from memory.
+    ClearTxs(*res);
+    return {};
+}
+
+util::Result<std::vector<CWallet::MapWalletIt>> CWallet::RemoveTxs(WalletBatch& batch, std::vector<uint256>& txs_to_remove) {
```
Comment (achow101) [inline src/wallet/wallet.cpp:2355]: In 4b3ec698a228fc531cca946f9a572e1dbdc4b0d5 "wallet: prepare RemoveTxs to be used within a broader db txn procedure"

I don't really like this refactor. This means that any callers of this particular overload will have to remember to do `ClearTxs`, whereas callers of the other one don't. This seems like something that could result in programming mistakes in the future. Generally our overloads should do basically the same things, just with the option of providing different arguments. Not clearing the txs from `mapWallet` in one overload is a pretty big difference that may not be easily noticed.

I think it would actually be a bit cleaner and better to have `RemoveTxs` detect if a database transaction has been started already and just choose to not begin a transaction if that's the case.

Comment (furszy) [reply src/wallet/wallet.cpp]: Ended up introducing a mechanism to listen for the db txn outcomes. So internal procedures can register callbacks to update the in-memory state only when the data is successfully committed to disk.

**Review by furszy — COMMENTED** [review]
Updated per feedback. Thanks achow.

**Comment by furszy** [comment]
rebased after #29130 merge.

**Review by furszy — COMMENTED** [review]
Rebased due a hidden conflict with #26606.

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -3696,25 +3696,17 @@ DescriptorScriptPubKeyMan& CWallet::SetupDescriptorScriptPubKeyMan(WalletBatch&
     return *out;
 }
 
-void CWallet::SetupDescriptorScriptPubKeyMans(const CExtKey& master_key)
+void CWallet::SetupDescriptorScriptPubKeyMans(WalletBatch& batch, const CExtKey& master_key)
```
Comment (achow101) [inline src/wallet/wallet.cpp:3699]: In 4e72821e1961b58559f49a4746bc8d15d116f6b7 "wallet: ApplyMigrationData, group all db txs into a single atomic operation"

This could be its own commit.

**Review by achow101 — COMMENTED** [review]
Concept ACK

I'm not a huge fan of all the boilerplate and wrappers that this ends up adding, but I also don't see a better way to do this.

In `ApplyMigrationData`, it looks like we're creating batchs for the watchonly and solvable wallets multiple times. Those could be consolidated in this PR as well.

File: src/wallet/wallet.cpp (hunk 8 above)
Comment (furszy) [reply src/wallet/wallet.cpp:3699]: Done

**Review by furszy — COMMENTED** [review]
> In ApplyMigrationData, it looks like we're creating batchs for the watchonly and solvable wallets multiple times. Those could be consolidated in this PR as well.

Done. Thanks.

File: src/wallet/wallet.cpp (hunk 8 above)
Comment (furszy) [reply src/wallet/wallet.cpp:3699]: Done

**Comment by achow101** [comment]
ACK ad1e3620af88f73b869b4a2dbb74e03f5cd93517

File: src/bench/wallet_migration.cpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2024 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#include <config/bitcoin-config.h> // IWYU pragma: keep
+
+#include <bench/bench.h>
+#include <interfaces/chain.h>
+#include <node/context.h>
+#include <test/util/mining.h>
```
Comment (HerWayBit) [inline src/bench/wallet_migration.cpp:10]: https://cash.app/$SelfMadeLady270

**Review by HerWayBit — APPROVED** [review]

File: src/Makefile.bench.include [hunk 10]
```diff
@@ -95,6 +95,7 @@ bench_bench_bitcoin_SOURCES += bench/wallet_create.cpp
 bench_bench_bitcoin_SOURCES += bench/wallet_loading.cpp
 bench_bench_bitcoin_SOURCES += bench/wallet_create_tx.cpp
 bench_bench_bitcoin_SOURCES += bench/wallet_ismine.cpp
+bench_bench_bitcoin_SOURCES += bench/wallet_migration.cpp
```
Comment (hebasto) [inline src/Makefile.bench.include:98]: Changes to the Autotools-based build system might be dropped altogether. Otherwise, a conflict is unavoidable once https://github.com/bitcoin/bitcoin/pull/30664 is merged, which is expected to happen tomorrow.

Comment (furszy) [reply src/Makefile.bench.include]: pushed.

**Review by theStack — COMMENTED** [review]
I'm not seeing any performance improvements locally with a non-debug build running on an SSD (benchmark shows ~13s both with and without the batching), but still think this PR is important for consistency reasons, i.e. to avoid the new wallet be left in a "half-migrated" state in case anything goes wrong. Changes look good to me at first glance (left only some nits), will do a deeper review round tomorrow.

Might be worth to update the PR description with benchmark results from release builds, as only those are relevant for end users and the speed-up could vary significantly?

File: src/bench/wallet_migration.cpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+#include <test/util/setup_common.h>
+#include <wallet/test/util.h>
+#include <wallet/context.h>
+#include <wallet/receive.h>
+#include <wallet/wallet.h>
+
+#include <optional>
+
+#ifdef USE_BDB // only enable benchmark when bdb and sqlite are enabled
+#ifdef USE_SQLITE
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/bench/wallet_migration.cpp]: done

File: src/bench/wallet_migration.cpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    bilingual_str error;
+    auto database = MakeWalletDatabase(fs::PathToString(test_setup->m_path_root / "legacy"), options, status, error);
+    uint64_t create_flags = 0;
+    auto wallet = TestLoadWallet(std::move(database), context, create_flags);
+
+    // Add watch-only addresses
+    std::vector<CScript> scripts_watch_only;
+    for (int w = 0; w < NUM_WATCH_ONLY_ADDR; ++w) {
+        CKey key;
+        key.MakeNewKey(true);
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/bench/wallet_migration.cpp]: done

File: src/wallet/wallet.cpp [hunk 13]
```diff
@@ -1701,10 +1701,16 @@ bool CWallet::CanGetAddresses(bool internal) const
 }
 
 void CWallet::SetWalletFlag(uint64_t flags)
+{
+    WalletBatch batch(GetDatabase());
+    return SetWalletFlagWithDB(batch, flags);
+}
+
+void CWallet::SetWalletFlagWithDB(WalletBatch& batch, uint64_t flags)
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/wallet.cpp:1709]: I think this was part of the first changes I did, I do like more the overloaded function approach as is less verbose without loosing readability. But later on I saw that we are using the `WithDB` suffix and changed the rest of the commits.
Will change `DeleteRecords` to use it. Thanks.

File: src/wallet/wallet.cpp [hunk 14]
```diff
@@ -3746,21 +3746,27 @@ void CWallet::SetupDescriptorScriptPubKeyMans(const CExtKey& master_key)
     if (!batch.TxnCommit()) throw std::runtime_error("Error: cannot commit db transaction for descriptors setup");
 }
 
+void CWallet::SetupOwnDescriptorScriptPubKeyMans()
+{
+    assert(!IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER));
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/wallet.cpp]: sure. done

File: src/wallet/wallet.cpp [hunk 15]
```diff
@@ -4104,12 +4099,13 @@ util::Result<void> CWallet::ApplyMigrationData(MigrationData& data)
...
     if (!IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+        WalletBatch local_wallet_batch(GetDatabase());
         // Use the existing master key if we have it
         if (data.master_key.key.IsValid()) {
-            SetupDescriptorScriptPubKeyMans(data.master_key);
+            SetupDescriptorScriptPubKeyMans(local_wallet_batch, data.master_key);
         } else {
             // Setup with a new seed if we don't.
-            SetupOwnDescriptorScriptPubKeyMans();
+            SetupOwnDescriptorScriptPubKeyMans(local_wallet_batch);
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/wallet.cpp]: > in [c436b51](https://github.com/bitcoin/bitcoin/commit/c436b516450b35bb01393559d935e1b807e5a835): shouldn't those calls also happen as a batch transaction, i.e. wrapped around a TxnBegin/TxnCommit pair (at least in the context of this commit)?

yeah, they should, but that would make this commit and the last one more verbose without much gain. Maybe it's better to leave it as is for simplicity?

**Comment by furszy** [comment]
Updated per feedback, thanks @theStack!

> I'm not seeing any performance improvements locally with a non-debug build running on an SSD (benchmark shows ~13s both with and without the batching), but still think this PR is important for consistency reasons, i.e. to avoid the new wallet be left in a "half-migrated" state in case anything goes wrong. Changes look good to me at first glance (left only some nits), will do a deeper review round tomorrow.
>
> Might be worth to update the PR description with benchmark results from release builds, as only those are relevant for end users and the speed-up could vary significantly?

Yeah, after all the other PRs got merged, this one is important for people running on HDDs. See https://github.com/bitcoin/bitcoin/pull/28037#issuecomment-1952211224, which shared a pretty nice time improvement. Going from a 30 minutes migration to one that took only 80 seconds.
Still, that was a long time ago. Will give it a new run on an HDD and update the PR description.

**Review by theStack — APPROVED** [review]
Code-review ACK a4e1ba722be5dbd71a3d21e5da82aee5409d22bd

Thanks for following up! 

> Yeah, now that all the other PRs got merged, things changed a bit. Aside from the consistency reasons, this PR is important for people running on HDDs. See https://github.com/bitcoin/bitcoin/pull/28037#issuecomment-1952211224, which shared a pretty nice time improvement. Going from a 30 minutes migration to one that took only 80 seconds.
Still, that was a long time ago. Will give it a new run on an HDD and update the PR description.

That sounds pretty impressive (>20x?). Would be very interesting indeed to re-benchmark with this branch again (cc @willcl-ark), though as said above the changes also make sense to me if the performance improvement is not that significant.

**Comment by furszy** [comment]
PR description updated. Created #31000 so we can all run the benchmark on an external HDD in a more friendly manner.
Have seen a 15% speedup locally on a USB 3.2 pendrive.

**Review by pablomartin4btc — COMMENTED** [review]
ACK a4e1ba722be5dbd71a3d21e5da82aee5409d22bd

<details>
<summary>I've ran the benchmarks locally on SSD and on an external USB 3.2 pendrive and got an improvement of ~46% (i7-1260P, 32GB RAM, Ubuntu 22.04 - using <code>pyperf system tune</code> to get stable results).</summary>

- This PR:

```
./build/src/bench/bench_bitcoin -filter=WalletMigration

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|    5,322,062,123.00 |                0.19 |    0.0% |      5.32 | `WalletMigration`
```

- `master`

```
./build/src/bench/bench_bitcoin -filter=WalletMigration

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|   11,362,482,640.00 |                0.09 |    0.0% |     11.36 | `WalletMigration`

```
</details>

**Comment by furszy** [comment]
@josibake @achow101 friendly ping. It seems we are quite close here.

**Comment by furszy** [comment]
Sad rebase due to tiny conflict with #30937. Ready to go.

**Comment by achow101** [comment]
ACK c98fc36d094a08d44f3c95431db2c5f34a96cc73

**Review by theStack — APPROVED** [review]
re-ACK c98fc36d094a08d44f3c95431db2c5f34a96cc73
(as per `$ git range-diff a4e1ba7...c98fc36`)

**Review by pablomartin4btc — APPROVED** [review]
re-ACK c98fc36d094a08d44f3c95431db2c5f34a96cc73

(no diff since my previous review, just rebased)
